src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -329,6 +331,35 @@ private Object readResolve() {
             return getDisplayName();
         }
 
+        @Restricted(NoExternalUse.class) // for Jelly
+        public @CheckForNull String getEnclosingLabel() throws Exception {
+            FlowNode executorStepNode = context.get(FlowNode.class);
+            if (executorStepNode != null) {
+                for (StepExecution exec : executorStepNode.getExecution().getCurrentExecutions(true).get(1, TimeUnit.MINUTES)) {
+                    FlowNode runningNode = exec.getContext().get(FlowNode.class);
+                    if (runningNode != null) {
+                        // See if this step is inside our node {} block, and track the associated label.
+                        boolean match = false;
+                        String label = null;
+                        FlowNodeSerialWalker.EnhancedIterator it = new FlowNodeSerialWalker(runningNode).iterator();
```

Comment:
> am I missing a set of full hard references?

At least for builds initiated during the current Jenkins session, the whole graph is in memory, via `FlowNode.parents`.